### PR TITLE
feat: add  axum compat

### PR DESCRIPTION
**Pull request by oknozor**
I am almost there, just need to figure out the digest header part and write an example.

I also took the opportunity to turn the federation example into an actual cargo `[[example]]` so it can depend on the actix feature flag. 

I will figure out the digest part soon. Meanwhile you might want to start reviewing :)

### CODE REVIEW

File: src/core/signatures.rs [hunk 1]
```diff
@@ -131,3 +94,55 @@ impl PublicKey {
...
+#[cfg(feature = "axum")]
+pub mod axum {
+    use axum::http::Request;
+
+    /// Verifies the HTTP signature on an incoming inbox request.
+    pub fn verify_signature<B>(
+        request: &Request<B>,
+        public_key: &str,
+    ) -> Result<(), anyhow::Error> {
+        todo!("Tower Http does not have this feature yet")
```
Comment (Nutomic) [inline src/core/signatures.rs:146]: You can directly use the [http-signature-normalization](https://crates.io/crates/http-signature-normalization) crate here, which should work independently of the web framework.

File: src/core/inbox.rs [hunk 2]
```diff
@@ -1,53 +1,124 @@
...
+            .await?;
+        verify_signature(&request, actor.public_key())?;
+
+        debug!("Verifying activity {}", activity.id().to_string());
+        activity.verify(data, request_counter).await?;
+
+        debug!("Receiving activity {}", activity.id().to_string());
+        activity.receive(data, request_counter).await?;
+        Ok(())
+    }
```
Comment (Nutomic) [inline src/core/inbox.rs:123]: Would be nice if you could extract common functionality into a helper function, so that its not duplicated in two places.

**Comment by MTRNord** [comment]
Thanks @oknozor :) Seems I was too slow :D Closed my PR in favor of this as it is already way ahead

**Comment by oknozor** [comment]
There is still a lot to do, the current Instance implementation is tightly coupled to actix runtile. 
I am looking for a way to abstract background-jobs-actix Manager. Suggestions are welcome @Nutomic :)

File: examples/federation-axum/instance.rs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    let user = ObjectId::<MyUser>::new(url)
+        .dereference_local(&data)
+        .await
+        .expect("Failed to dereference user")
+        .into_apub(&data)
+        .await
+        .expect("Failed to convert to apub user");
+
+    let json =
```
Comment (MTRNord) [inline examples/federation-axum/instance.rs:139]: Not sure if more or less elegant but I did add myself this small wrapper for doing this:

```
#[derive(Debug, Clone, Copy, Default)]
struct ApubJson<Json: Serialize>(Json);

impl<Json: Serialize> IntoResponse for ApubJson<Json> {
    fn into_response(self) -> axum::response::Response {
        let mut response = axum::response::Json(self.0).into_response();
        response.headers_mut().insert(
            header::CONTENT_TYPE,
            APUB_JSON_CONTENT_TYPE.parse().unwrap(),
        );
        response
    }
}
```

The benefit is that instead of `impl IntoResponse` I can return `Result<ApubJson<Person>, ServerError>` which makes it possible to use typed errors :)

Maybe thats something that could be upstreamed into this PR?

**Comment by Nutomic** [comment]
> I am looking for a way to abstract background-jobs-actix Manager. Suggestions are welcome @Nutomic :)

This one doesnt depend on `actix-web` (web framework) but on [actix-rt](https://crates.io/crates/actix-rt) which is an async runtime. There should be no problem to use it with axum.

File: Cargo.toml [hunk 4]
```diff
@@ -30,23 +30,43 @@ dyn-clone = "1.0.9"
...
 http-signature-normalization-actix = { version = "0.6.1", default-features = false, features = ["server", "sha-2"], optional = true }
 axum = { version = "0.6.0-rc.5", features = ["json", "headers", "macros"], optional = true }
 
+# Axum
+tower-http = { version = "0.3", features = ["map-request-body", "util"], optional = true }
+tower = { version = "0.4.13", optional = true }
+hyper = { version = "0.14", optional = true }
+
+## FIXME: this is for debugging purpose
+axum-macros = "0.3.0"
```
Comment (Nutomic) [inline Cargo.toml:45]: Just make it a dev dependency then?

File: src/core/inbox.rs [hunk 5]
```diff
@@ -93,13 +95,9 @@ mod axum_imp {
             + From<Error>
             + From<<ActorT as ApubObject>::Error>
             + From<serde_json::Error>,
-        // FIXME
         // + From<http_signature_normalization_actix::digest::middleware::VerifyError>,
         <ActorT as ApubObject>::Error: From<Error> + From<anyhow::Error>,
     {
-        // ensure that payload hash was checked against digest header by middleware
-        /// FIXME
-        // DigestVerified::from_request(&request, &mut Payload::None).await?;
```
Comment (Nutomic) [inline src/core/inbox.rs]: This is only necessary in the actix code to avoid security problems in case a library user forgets to use the actix middleware for signature verification. It would be better to make the signature check directly in the library, and then remove the actix VerifyDigest middleware as well as DigestVerified

File: examples/federation-axum/utils.rs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+/// Just generate random url as object id. In a real project, you probably want to use
+/// an url which contains the database id for easy retrieval (or store the random id in db).
+pub fn generate_object_id(hostname: &str) -> Result<Url, ParseError> {
+    let id: String = thread_rng()
+        .sample_iter(&Alphanumeric)
+        .take(7)
+        .map(char::from)
+        .collect();
+    Url::parse(&format!("http://{}/objects/{}", hostname, id))
+}
```
Comment (Nutomic) [inline examples/federation-axum/utils.rs:13]: Can move this into some shared file (or even into the library itself?)

Comment (oknozor) [reply examples/federation-axum/utils.rs:13]: Dw I will get into refactoring once it works

**Comment by oknozor** [comment]
This is going to be tricky to implement. 
Axum requires futures that are `Send` 
```error: future cannot be sent between threads safely
   --> examples/federation-axum/instance.rs:121:1
    |
121 | / /// Handles requests to fetch user json over HTTP
122 | | async fn http_get_user(
123 | |     State(data): State<InstanceHandle>,
124 | |     request: Request<Body>,
...   |
146 | |         .expect("failed to build response")
147 | | }
    | |_^ future returned by `http_get_user` is not `Send`
    |
    = help: the trait `Send` is not implemented for `dyn Future<Output = Result<std::option::Option<MyUser>, error::Error>>`
note: required by a bound in `__axum_macros_check_http_get_user_future::check`
   --> examples/federation-axum/instance.rs:121:1
    |
121 | / /// Handles requests to fetch user json over HTTP
122 | | async fn http_get_user(
123 | |     State(data): State<InstanceHandle>,
124 | |     request: Request<Body>,
...   |
146 | |         .expect("failed to build response")
147 | | }
    | |_^ required by this bound in `__axum_macros_check_http_get_user_future::check`

**Comment by oknozor** [comment]
I made everything `Send + Sync` so now it compile both on actix and axum. 
However since we have no actix runtime on the axum example it fails at runtime trying to create the activity queue. 

I guess I will have to implement some queue mechanism on the tokio runtime ? 
Or maybe we can use something like this  https://crates.io/crates/rusty-sidekiq

What do you thing @Nutomic 

Btw It's a bit of a mess right now but I will factorize everything as soon as I get it work on both actix and axum.

**Comment by MTRNord** [comment]
> I made everything `Send + Sync` so now it compile both on actix and axum. However since we have no actix runtime on the axum example it fails at runtime trying to create the activity queue:
> 
> ```
> thread 'main' panicked at 'System is not running', /home/okno/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-rt-2.7.0/src/system.rs:120:21
> ```
> 
> I guess I will have to implement some queue mechanism on the tokio runtime ? Or maybe we can use something like this https://crates.io/crates/rusty-sidekiq

Are you sure that it doesnt work if you just add https://docs.rs/actix-rt/latest/actix_rt/attr.main.html on the main function instead of using tokio's main macro? (Iirc actix uses tokio internally for this in a single thread runtime. Though that kinda would defeat the purpose of axum arguably because at least my reason to use it over actix-web was that it is capable of full multithreading)

**Comment by MTRNord** [comment]
> > I made everything `Send + Sync` so now it compile both on actix and axum. However since we have no actix runtime on the axum example it fails at runtime trying to create the activity queue:
> > ```
> > thread 'main' panicked at 'System is not running', /home/okno/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-rt-2.7.0/src/system.rs:120:21
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I guess I will have to implement some queue mechanism on the tokio runtime ? Or maybe we can use something like this https://crates.io/crates/rusty-sidekiq
> 
> Are you sure that it doesnt work if you just add https://docs.rs/actix-rt/latest/actix_rt/attr.main.html on the main function instead of using tokio's main macro? (Iirc actix uses tokio internally for this in a single thread runtime. Though that kinda would defeat the purpose of axum arguably because at least my reason to use it over actix-web was that it is capable of full multithreading)

After trying this it seems to be running:

image

File: examples/federation-axum/instance.rs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+                    .layer(middleware::from_fn(verify_request_payload)),
+            )
+            .route("/objects/:user_name", get(http_get_user))
+            .with_state(instance);
+
+        // run it
+        let addr = SocketAddr::from_str(hostname)?;
+        let server = axum::Server::bind(&addr).serve(app.into_make_service());
+
+        task::spawn(server);
```
Comment (MTRNord) [inline examples/federation-axum/instance.rs:113]: This probably should have a comment that in reallife scenarios this should not be on a separate spawn thread as otherwise the app would just exit instead of waiting :)

Comment (oknozor) [reply examples/federation-axum/instance.rs:116]: It's not exiting, there was some request part missing on the signature verify logic. 
This is fixed now.

File: examples/federation-axum/utils.rs (hunk 6 above)
Comment (oknozor) [reply examples/federation-axum/utils.rs:13]: I am not sure this could be factorized, the `ApubObject` implementation requires to have the concrete `type DataType = InstanceHandle;` type. 
Having generics in the examples would be confusing imo.

File: src/core/axum/json.rs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+/// ```rust, no_run
+/// use activitypub_federation::deser::context::WithContext;
+///  async fn http_get_user() -> Result<ApubJson<WithContext<Person>>, Error> {
+///     let user = WithContext::new_default(M);
+///
+///     Ok(ApubJson(WithContext::new_default(MyUser::default())))
+/// }
+/// ```
+#[derive(Debug, Clone, Copy, Default)]
+pub struct ApubJson<Json: Serialize>(Json);
```
Comment (MTRNord) [inline src/core/axum/json.rs:19]: The inner Json must also be public, otherwise it cannot be set from outside this crate.

**Comment by oknozor** [comment]
Here we are ! 

The last thing to figure out is how to get rid of `background-jobs`, at least for the axum implementation. 

Currently it works but only with actix runtime. Maybe we can merge this as is and fix it later. 
This would require a note in the readme since its a bit counter intuitive to use actix runtime with axum. 

I am planning to take a closer look at worker queue management tomorrow to fix  I already had a quick look at [rusty-sidekiq](https://crates.io/crates/rusty-sidekiq) the only problem is that it require a redis running instance. 

I am not sure you want this. at this point I am tempted to fork and make a `rusty-sidekiq` compatible version of this crate. The alternative would be to implement something like `background-jobs` running on tokio. 

Let me now what you think @Nutomic. 

PS: do you prefer that I squash all commits into a single one or rename the existing one so they are clearer ? 


Cheers

**Comment by Nutomic** [comment]
The part about `?Send` in async trait seems like bad copy-paste on my part. After removing it from Lemmy trait definitions, it compiles just fine with this PR. 

About `background-jobs`, wouldnt it be enough initialize actix runtime inside the crate if necessary, so that axum users dont have to do it manually? Then there would be nothing counterintuitive.

In prinicple Im not against using another crate for the job queue, but it should be one thats already proven in production, and it should have clear technical advantages. I dont want to add an external dependency like Redis, as it makes server administration more complicated for no benefit. 

[background-jobs](https://git.asonix.dog/asonix/background-jobs) also allows usage with different backends, so that might be an option. You can also get in contact with the author asonix, he is always very helpful.

Squasing doesnt matter, i can easily do that with a button click while merging.

**Comment by MTRNord** [comment]
> About background-jobs, wouldnt it be enough initialize actix runtime inside the crate if necessary, so that axum users dont have to do it manually? Then there would be nothing counterintuitive.

I think this might cause issues if the app using it has a tokio runtime. Tokio tends to not like having runtimes nested and actix-rt just wraps a tokio runtime. I guess best is to just try it though.

File: src/core/actix/inbox.rs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    for<'de2> <ActorT as ApubObject>::ApubType: serde::Deserialize<'de2>,
+    <Activity as ActivityHandler>::Error: From<anyhow::Error>
+        + From<Error>
+        + From<<ActorT as ApubObject>::Error>
+        + From<serde_json::Error>
+        + From<http_signature_normalization_actix::digest::middleware::VerifyError>,
+    <ActorT as ApubObject>::Error: From<Error> + From<anyhow::Error>,
+{
+    // ensure that payload hash was checked against digest header by middleware
+    DigestVerified::from_request(&request, &mut Payload::None).await?;
```
Comment (Nutomic) [inline src/core/actix/inbox.rs:34]: This line isnt can be removed now as we are verifying the digest directly in the lib now. Maybe `http-signature-normalization-actix` dependency can also be removed then.

Comment (MTRNord) [reply src/core/actix/inbox.rs:34]: Removing it would be great due to its licensing. That dependency is very restrictive about the license. (freedom of licenses of they are compatible would be nice but they limit it to 2 choices)

**Comment by Nutomic** [comment]
The code looks good to me, except for that minor change mentioned above. Also it seems that Lemmy works with these changes without major adjustments ([see pr](https://drone.join-lemmy.org/LemmyNet/lemmy/1717)). If @MTRNord doesnt have any more comments then i can merge it.

Do you want me to make a `0.4.0` release then? Otherwise you could also develop using git dependency now, and I make a release later in case more breaking changes are needed.

```
activitypub_federation = { git = "https://github.com/LemmyNet/activitypub-federation-rust.git", features = ["axum"] } 
```

**Comment by MTRNord** [comment]
I think I already added all comments I had yesterday. I also am using it already locally and the example seems to work. I didn't do any extra tests beyond this since I am new to activitypub but so far it seems like there are no obvious bugs or issues with it.

Tldr: no additions from my side.

**Comment by oknozor** [comment]
Updated as suggested, also the `verify_digest` method is now generic and used for both actix and axum.
